sdk/python/lib/pulumi/runtime/known_types.py

```diff
@@ -12,294 +12,114 @@
 to reference types that are defined there.
 
 In order to break this circular reference, and to be clear about what types
-the runtime knows about and treats specially, this module exports a number of
-"known type" decorators that can be applied to types in `pulumi` to indicate
-that they are specially treated.
+the runtime knows about and treats specially, we defer loading of the types from
+within the functions themselves.
 
-The implementation of this mechanism is that, for every known type, that type
-is stashed away in a global variable. Whenever the runtime wants to do a type
-test using that type (or instantiate an instance of this type), it uses the
-functions defined in this module to do so.
+This implementation allows for overriding types used as stubs for testing
+(see test/test_next_serialize.py)
 """
 from typing import Any, Optional
 
 _custom_resource_type: Optional[type] = None
-"""The type of CustomResource. Filled-in as the Pulumi package is initializing."""
+"""The type of CustomResource."""
 
 _custom_timeouts_type: Optional[type] = None
-"""The type of CustomTimeouts. Filled-in as the Pulumi package is initializing."""
+"""The type of CustomTimeouts."""
 
 _asset_resource_type: Optional[type] = None
-"""The type of Asset. Filled-in as the Pulumi package is initializing."""
+"""The type of Asset."""
 
 _file_asset_resource_type: Optional[type] = None
-"""The type of FileAsset. Filled-in as the Pulumi package is initializing."""
+"""The type of FileAsset."""
 
 _string_asset_resource_type: Optional[type] = None
-"""The type of StringAsset. Filled-in as the Pulumi package is initializing."""
+"""The type of StringAsset."""
 
 _remote_asset_resource_type: Optional[type] = None
-"""The type of RemoteAsset. Filled-in as the Pulumi package is initializing."""
+"""The type of RemoteAsset."""
 
 _archive_resource_type: Optional[type] = None
-"""The type of Archive. Filled-in as the Pulumi package is initializing."""
+"""The type of Archive."""
 
 _asset_archive_resource_type: Optional[type] = None
-"""The type of AssetArchive. Filled-in as the Pulumi package is initializing."""
+"""The type of AssetArchive."""
 
 _file_archive_resource_type: Optional[type] = None
-"""The type of FileArchive. Filled-in as the Pulumi package is initializing."""
+"""The type of FileArchive."""
 
 _remote_archive_resource_type: Optional[type] = None
-"""The type of RemoteArchive. Filled-in as the Pulumi package is initializing."""
+"""The type of RemoteArchive."""
 
 _stack_resource_type: Optional[type] = None
-"""The type of Stack. Filled-in as the Pulumi package is initializing."""
+"""The type of Stack."""
 
 _output_type: Optional[type] = None
-"""The type of Output. Filled-in as the Pulumi package is initializing."""
+"""The type of Output."""
 
 _unknown_type: Optional[type] = None
-"""The type of unknown. Filled-in as the Pulumi package is initializing."""
-
-def asset(class_obj: type) -> type:
```

Comment:
Alternatively, adding stronger typing to these may have helped, but I haven't tried it, so maybe not...

```python
if TYPE_CHECKING:
     from .. import Asset

def asset(class_obj: Type['Asset']) -> Type['Asset']:
```

But, I'm in favor of getting rid of the `known_types` decorators, if we can.

Reply:
I tried this but it doesn't work all the way. It does help with the type-hinting (it'll show that you're missing a required argument) but it doesn't show the docstrings.
  
<img width="441" alt="Screen Shot 2020-07-08 at 11 42 08 AM" src="https://user-images.githubusercontent.com/24326455/86957799-1558cc80-c110-11ea-82bb-51bf0416021e.png">

And yeah I think the `known_types` decorators are an interesting but non-standard workaround for this particular issue so worth removing for that reason as well.